Source/CesiumRuntime/Private/Tests/SubLevels.spec.cpp

```diff
@@ -0,0 +1,338 @@
+
+    pGeoreference = nullptr;
+    for (TActorIterator<ACesiumGeoreference> it(pWorld); it; ++it) {
+      pGeoreference = *it;
+    }
+
+    trackForPlay(pGeoreference);
+
+    pPawn = pWorld->SpawnActor<AGlobeAwareDefaultPawn>();
+    trackForPlay(pPawn);
+    pPawn->AutoPossessPlayer = EAutoReceiveInput::Player0;
+  });
+
+  AfterEach([this]() {
+    pSubLevel1->SetIsTemporarilyHiddenInEditor(true);
+    pSubLevel2->SetIsTemporarilyHiddenInEditor(true);
+  });
+
+  It("initially hides sub-levels in the Editor", [this]() {
+    TestTrue("pGeoreference is valid", IsValid(pGeoreference));
+    TestTrue("pSubLevel1 is valid", IsValid(pSubLevel1));
+    TestTrue("pSubLevel2 is valid", IsValid(pSubLevel2));
+    TestTrue(
+        "pSubLevel1 is hidden",
+        pSubLevel1->IsTemporarilyHiddenInEditor(true));
+    TestTrue(
+        "pSubLevel2 is hidden",
+        pSubLevel1->IsTemporarilyHiddenInEditor(true));
+  });
+
+  Describe(
+      "copies CesiumGeoreference origin changes to the active sub-level in the Editor",
+      [this]() {
+        BeforeEach(EAsyncExecution::TaskGraphMainThread, [this]() {
+          pSubLevel1->SetIsTemporarilyHiddenInEditor(false);
+        });
+        It("", EAsyncExecution::TaskGraphMainThread, [this]() {
+          TestFalse(
+              "pSubLevel1 is hidden",
+              pSubLevel1->IsTemporarilyHiddenInEditor(true));
+
+          pGeoreference->SetGeoreferenceOriginLongitudeLatitudeHeight(
+              FVector(1.0, 2.0, 3.0));
+          TestEqual("Longitude", pLevelComponent1->GetOriginLongitude(), 1.0);
+          TestEqual("Latitude", pLevelComponent1->GetOriginLatitude(), 2.0);
+          TestEqual("Height", pLevelComponent1->GetOriginHeight(), 3.0);
+        });
+      });
+
+  Describe(
+      "copies active sub-level origin changes to the CesiumGeoreference in the Editor",
+      [this]() {
+        BeforeEach(EAsyncExecution::TaskGraphMainThread, [this]() {
+          pSubLevel1->SetIsTemporarilyHiddenInEditor(false);
+        });
+        It("", EAsyncExecution::TaskGraphMainThread, [this]() {
+          TestFalse(
+              "pSubLevel1 is hidden",
+              pSubLevel1->IsTemporarilyHiddenInEditor(true));
+
+          pLevelComponent1->SetOriginLongitudeLatitudeHeight(
+              FVector(4.0, 5.0, 6.0));
+          TestEqual("Longitude", pGeoreference->OriginLongitude, 4.0);
+          TestEqual("Latitude", pGeoreference->OriginLatitude, 5.0);
+          TestEqual("Height", pGeoreference->OriginHeight, 6.0);
+        });
+      });
+
+  Describe(
+      "does not copy inactive sub-level origin changes to the CesiumGeoreference in the Editor",
+      [this]() {
+        BeforeEach(EAsyncExecution::TaskGraphMainThread, [this]() {
+          pSubLevel1->SetIsTemporarilyHiddenInEditor(false);
+        });
+        BeforeEach(EAsyncExecution::TaskGraphMainThread, [this]() {
+          pSubLevel1->SetIsTemporarilyHiddenInEditor(true);
+        });
```

Comment:
Is there an async-related reason that these are in separate clauses? I'm probably missing something

Reply:
It's not strictly necessary, but it's a slighter better test this way. Basically this means:

1. Set the sub-level to be active, and then wait a frame so it actually becomes active.
2. Set the sub-level to be inactive, and wait a frame so it actually becomes inactive.
3. Make sure that the previously-active (but now no longer active) level isn't affected by the georeference origin change.

If we didn't activate it first, we'd instead be testing if a sub-level that was never active is affected by a georeference origin change. And that's somewhat less likely to ever be a problem.

Reply:
I added some comments to make this clearer.